Add support for custom pam configuration

dmulder opened the pull request:
Fixes #32 

The goal of this MR is to be able to inject custom pam configuration using `pam-config`. We'll utilize configuration files which are similar in principal to Ubuntu's PAMConfig files, like this:

```
Name: pam_testdummy.so
Auth:
	auth	sufficient    pam_testdummy.so    ignore_unknown_user
Auth-Priority: 1500
Account:
	account	required      pam_testdummy.so    ignore_unknown_user
Account-Priority: 750
Password:
	password	sufficient    pam_testdummy.so    use_authtok
Password-Priority: 1200
Session:
	session	optional      pam_testdummy.so
Session-Priority: 2100
```

To start with, one major difference will be that the `Priority` value will need to be provided for each type. Default module order is defined in `src/module_priorities.h` (this mirrors the existing behavior). Custom config files will override the built-in modules by name and priority (for example, a custom module called `sss` would override the built-in one, and will utilize the priorities specified in the config).
We'll need to generate module cli options depending on the detection of custom configs (so that we can make the process seamless).

dmulder committed:
Give pam modules priorities

Later this will be modified to allow custom
modules to be inserted in the module lists.

Signed-off-by: David Mulder <dmulder@samba.org>

dmulder commented on src/module_priorities.h:
```diff
@@ -0,0 +1,106 @@
+#define PAM_NAM_SESSION_PRIORITY 1500
+#define PAM_UMASK_SESSION_PRIORITY 1600
+#define PAM_SSH_SESSION_PRIORITY 1700
+#define PAM_GNOME_KEYRING_SESSION_PRIORITY 1900
```
FYI, there is a gap here, since the existing module order actually has a typo. The pam_selinux module is listed twice, in first place and 18th place.

dmulder committed:
Dynamicaly load pam modules from configuration

Signed-off-by: David Mulder <dmulder@samba.org>

dmulder commented:
@thkukuk this is still pretty rough around the edges, and doesn't yet handle the case where a custom config is defined which overwrites a built-in one, but maybe you'd like to review the approach so far?

dmulder committed:
Cause custom configs to override built-ins

Signed-off-by: David Mulder <dmulder@samba.org>

dmulder committed:
Ensure we don't duplicate custom configs

Signed-off-by: David Mulder <dmulder@samba.org>

dmulder commented:
> I hope to find time to look at this in the next days, but could be very well that I will only manage that in October.

That's understandable. There is no urgency, of course.

thkukuk commented:
I'm still traveling...

dmulder committed:
Replace libinih with libeconf

Signed-off-by: David Mulder <dmulder@samba.org>

dmulder committed:
Permit module behavior fallback

Signed-off-by: David Mulder <dmulder@samba.org>

dmulder committed:
Only free configurable pam modules

Signed-off-by: David Mulder <dmulder@samba.org>

dmulder commented:
> I also don't like the `__attribute__((constructor))` functions, I think this functions should be called by `main()`.
> Else it would make it unnecessary hard for others to understand how this works, as they have to search first for such functions.

I tried removing the `constructor` approach, but this doesn't work. I think it's because `common_module_list` needs to be loaded much earlier than the calling of `main`. We could still make this happen, but I think it will require some major restructuring of pam-config.

dmulder commented:
I've discovered a problem. The `write_config_configurable` function gets called for **every** configurable pam module that's discovered in the config files, writing the output for every one of those configurable modules, whenever any single configurable module is called for `--add`.
I'm not sure why.

dmulder committed:
Ensure each configurable module uses it's own opt_sets

Signed-off-by: David Mulder <dmulder@samba.org>

dmulder commented:
Ok, I resolved that problem by generating opt_sets for each configurable module, instead of using the same global static options for all the modules.

@thkukuk this is ready for review again! The only thing I couldn't rework was your request to remove the `__attribute__((constructor))`.

dmulder committed:
Ensure we call fallback stacks

Signed-off-by: David Mulder <dmulder@samba.org>

thkukuk commented:
Sorry for the delay again, but looks good to me.
Now we only need a little piece of documentation of the format in the manual page and removing the draft state, after that I'm fine to merge it.

dmulder commented:
> Sorry for the delay again, but looks good to me. Now we only need a little piece of documentation of the format in the manual page and removing the draft state, after that I'm fine to merge it.

There are still some bugs I was trying to work out. I'll work on this later, and update the man page.

thkukuk commented:
The next step for me will be to convert the autotool stuff to meson and add a better test suite...

dmulder committed:
Enable the configurable fallback when called

We have to manually enable a fallback module
prior to calling it, otherwise the fallback will
never print anything.

Signed-off-by: David Mulder <dmulder@samba.org>

dmulder commented:
[9d551bc](https://github.com/SUSE/pam-config/pull/33/commits/9d551bce085aff0a1562388546eeffbbd6aea891) should resolve the last remaining issue. Now I'll add something to the man page.

dmulder committed:
Explain custom pam-config modules in the man page

Signed-off-by: David Mulder <dmulder@samba.org>

dmulder commented:
@thkukuk finished!

thkukuk commented:
Ok, let's see what openQA says to this.

Thanks!
